src/Nethermind/Nethermind.Network/ForkInfo.cs

```diff
@@ -47,19 +47,28 @@ public static byte[] CalculateForkHash(ISpecProvider specProvider, long headNumb
             return forkHash;
         }
 
-        public static ForkId CalculateForkId(ISpecProvider specProvider, long headNumber, Keccak genesisHash)
+        public static ForkId CalculateForkId(ISpecProvider specProvider, long headNumber, ulong headTimestamp, Keccak genesisHash)
         {
-            byte[] forkHash = CalculateForkHash(specProvider, headNumber, genesisHash);
-            long next = 0;
-            for (int i = 0; i < specProvider.TransitionBlocks.Length; i++)
+
+            byte[] forkHash = CalculateForkHash(specProvider, headNumber, headTimestamp, genesisHash);
+            ulong next = 0;
+            ForkActivation[] transitionBlocks = specProvider.TransitionBlocks;
+            for (int i = 0; i < transitionBlocks.Length; i++)
             {
-                if (specProvider.TransitionBlocks[i] >= ImpossibleBlockNumberWithSpaceForImpossibleForks)
+                bool useTimestamp = false;
+                if (transitionBlocks[i].Timestamp != null)
+                    useTimestamp = true;
+                ulong transition = transitionBlocks[i].Timestamp ?? (ulong)transitionBlocks[i].BlockNumber;
+
+
+                if ((useTimestamp && transition >= ImpossibleTimestampWithSpaceForImpossibleForks)
+                    || (!useTimestamp && transition >= ImpossibleBlockNumberWithSpaceForImpossibleForks))
                 {
                     continue;
                 }
 
-                long transition = specProvider.TransitionBlocks[i].BlockNumber;
-                if (transition > headNumber)
+                if ((useTimestamp && transition > headTimestamp)
+                    || (!useTimestamp && transition > (ulong)headNumber))
```

Comment:
just make one if, will be more readable.

if (transitionBlocks[i].Timestamp is not null)
else

Reply:
Did not really find a much cleaner way than this.